Bug 2026387: Handle certificate rotation in pkg/metrics/server.go

dagrayvid opened the pull request:
This  PR is intended to solve RHBZ#2026387. 

With this change, a file watcher is added to the metrics server using the fsnotify package, to watch for changes to the TLS certificate for the metrics endpoint. I took some inspiration from https://github.com/openshift/cluster-ingress-operator/pull/334/. Instead of restarting the operator when the certificate changes as is done in that PR, the server will be shut down and restarted.

These changes have been tested by editing the node-tuning-operator-tls secret (comments can be added to the certificate / key by putting them outside of the `-----BEGIN CERTIFICATE-----` and `-----END CERTIFICATE-----`   lines) and by deleting the secret altogether which results in a new key pair being created.

After some review, I will do more testing using the same NTP process described in the BZ to "expire" the certificate.

@jmencak PTAL

Signed-off-by: David Gray <dagray@redhat.com>

fcami commented on pkg/metrics/server.go:
```diff
@@ -1,25 +1,32 @@
-	tlsKey = "/etc/secrets/tls.key"
+	certsDir = "/etc/secrets"
+	tlsCRT   = certsDir + "/tls.crt"
+	tlsKey   = certsDir + "/tls.key"
```
Could you use filepath.Join instead ( https://pkg.go.dev/path/filepath#Join )?

dagrayvid replied:
I had this thought, however we can't use a function call to set a constant.

fcami commented on pkg/metrics/server.go:
```diff
@@ -37,14 +44,77 @@ func RunServer(port int, stopCh <-chan struct{}) {
+	if err := srv.Shutdown(ctx); err != nil {
+		klog.Errorf("error or timeout stopping metrics listener: %v", err)
+	} else {
+		// TODO delete debugging lines
```
I think that comment is obsolete.

jmencak replied:
I think this line and the line below is the debugging that should be removed.  Anyways, still WiP and will be cleaned I believe.

fcami commented on pkg/metrics/server.go:
```diff
@@ -78,3 +148,66 @@ func Degraded(deg bool) {
+	}
+	defer file.Close()
+
+	hash := md5.New()
```
Could you use something "better" than MD5, as we're trying to minimizeMD5 usage in RHEL? sha256 for instance.

dagrayvid replied:
I had sha256 originally. I (perhaps naively) switched to md5 for the slight performance improvement because I figured security was not a concern here. I switched it back as performance really shouldn't be a problem here, and it is the current standard.

fcami commented on pkg/metrics/server.go:
```diff
@@ -1,25 +1,32 @@
-	tlsCRT = "/etc/secrets/tls.crt"
-	tlsKey = "/etc/secrets/tls.key"
+	certsDir = "/etc/secrets"
+	tlsCRT   = certsDir + "/tls.crt"
```
Minor nitpick: tlsCert seems more readable, and more aligned with tlsKey, than tlsCRT.

dagrayvid replied:
Agreed, I had the same thought myself.

dagrayvid replied:
Done

fcami reviewed (changes requested):
The change LGTM, 
The one concern I have is the behavior of the fsnotify watcher for the /etc/secrets directory. Is the secrets volume unmounted/mounted again on certificate renewal or are the certificate and key files deleted and then recreated inside the same volume? The umount/mount operation might throw the watcher off.
Is there a way to add a CI test for this, deleting the secrets twice?

jmencak reviewed (commented):
Thank you for the PR, David, much appreciated.  Overall this looks sane to me.  I didn't test the code, but I agree with François that a CI test for this would be useful.  Left some comments for inspiration/cleanup.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,14 +44,77 @@ func RunServer(port int, stopCh <-chan struct{}) {
+			stopServer(srv)
+		case event, ok := <-watcher.Events:
+			if !ok {
+				klog.Errorf("file watcher events channel closed")
```
I think this should never happen. My first thought is to just `continue` the loop, which will mean certificate rotation is disabled (no more events from watcher channel). 

Alternatively we could kill the whole operator in the case of unexpected failures like this. Any guidance you can provide based on how other parts of the operator handle similar situations?

jmencak replied:
> I think this should never happen. My first thought is to just `continue` the loop, which will mean certificate rotation is disabled (no more events from watcher channel).

As mentioned elsewhere, you're probably right the likelihood of this happening is very low.  The rest is dependent on the underlying `fsnotify` implementation which I have not studied... 

> 
> Alternatively we could kill the whole operator in the case of unexpected failures like this. Any guidance you can provide based on how other parts of the operator handle similar situations?

Something like `klog.Fatal()` (and let the operator be restarted) is certainly an option, but probably not here.  In the operator, we don't have any "retry" loops when things go really bad.  We do have one in the operand though.  Saying that, it is probably pointless to have one in the operator unless you just want to make a local one but I do not believe this is needed.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -78,3 +148,66 @@ func Degraded(deg bool) {
+
+	return hash.Sum(nil)
+}
+
```
Added, thanks.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -78,3 +148,66 @@ func Degraded(deg bool) {
+	return false
+}
+
+//Compute the md5 checksum for a file.
```
Fixed, thanks.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -78,3 +148,66 @@ func Degraded(deg bool) {
 	degradedState.Set(0)
 }
+
+// Determine if the certificates have changed and need to be updated
```
Fixed these, thanks.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -78,3 +148,66 @@ func Degraded(deg bool) {
+		klog.Infof("checksums before: %x, %x. checksums after: %x, %x", origCertMD5, origKeyMD5, currentCertMD5, currentKeyMD5)
+		// Check if the non-empty certificate / key files have actually changed
+		if !bytes.Equal(origCertMD5, currentCertMD5) && !bytes.Equal(origKeyMD5, currentKeyMD5) {
+			klog.Infof("cert and key changed, we should restart the server.")
```
Fixed that, thanks... I think I wrote this before I had implemented the restart ;)

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -78,3 +148,66 @@ func Degraded(deg bool) {
+		// One of the files is missing despite some file event
+		klog.Infof("certificate or key is missing or empty. Don't rotate.")
+		return false
+	} else {
```
Good catch, remove this nesting makes the function a lot cleaner.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -78,3 +148,66 @@ func Degraded(deg bool) {
+		}
+	}
+
+	klog.Infof("At least one of the cert / key files has not changed")
```
This was really just here for debugging purposes, I removed this one and changed a few other logs to V(1) or V(2).

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,14 +44,77 @@ func RunServer(port int, stopCh <-chan struct{}) {
+	}
+	defer watcher.Close()
+	if err := watcher.Add(certsDir); err != nil {
+		klog.Errorf("failed to add %v to watcher: %v", certsDir, err)
```
Good idea. I added this.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,14 +44,77 @@ func RunServer(port int, stopCh <-chan struct{}) {
+	if err := watcher.Add(certsDir); err != nil {
+		klog.Errorf("failed to add %v to watcher: %v", certsDir, err)
+	} else {
+		klog.Infof("watching: %v", certsDir)
```
Really just there for debugging. I removed this one.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,14 +44,77 @@ func RunServer(port int, stopCh <-chan struct{}) {
+				klog.Errorf("file watcher events channel closed")
+			}
+
+			klog.Infof("file: %v, event: %v", event.Name, event.Op)
```
Good catch. I made this V(2) as it may be useful for future debugging. I also made it so that we will skip the checksum step if the file event is chmod or remove. The other event types are create, write, and rename.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,14 +44,77 @@ func RunServer(port int, stopCh <-chan struct{}) {
-	<-stopCh
-	if err := srv.Shutdown(context.Background()); err != nil && err != http.ErrServerClosed {
-		klog.Errorf("error stopping metrics listener: %v", err)
+
```
Fixed.

dagrayvid commented:
Added a new revision with many of the comments resolved. I added it in a new commit just to make things clear, but I can squash the commits once we are done iterating.

I decided to move functions PodLabelsUsed,  ProfileCalculated, RegisterVersion and Degraded to metrics.go rather than server.go to keep server.go specific to running the server. To me this is a more logical separation, but open to other ideas.

jmencak commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
+	if err := srv.Shutdown(ctx); err != nil {
+		klog.Errorf("error or timeout stopping metrics listener: %v", err)
+	}
+	klog.V(2).Infof("stopping metrics server without error")
```
stopping or stopped?  This could be my English though, so you're in a better position to asses which is better.

dagrayvid replied:
You're definitely right this is awkward as written... It was fixed on my local copy :)

jmencak commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
+		case <-stopCh:
+			stopServer(srv)
+		case event, ok := <-watcher.Events:
+			if !ok {
```
I wonder if we even need this check.  I have been happily runing with the detection of channel closing in `tuned` for years.  Also, my advice of add "cert/key rotation will be disabled" is probably not a good one looking more closely.  My appologies.  This might be just a temporary blip and the rotation code might be completely unaffected.  `watcher.Add()` are fatal though in my view and not checking for `nil` will bring down the whole operator so we need to be extra careful there.

dagrayvid replied:
I removed this check. If somehow `event` were set to nil here, I believe the worst case is that we would check if the files really changed.

jmencak reviewed (commented):
Thank you for the changes, David.  Steadily making improvements.  My main concern now is error checking and handling `SIGTERM` and shutdown gracefully.  This needs to be tested.

jmencak commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			stopServer(srv)
```
After a little testing, this should not block `SIGTERM` even if you do not return.  However, at exit you'll have running goroutines you don't need to have.  You can check this with `runtime.NumGoroutine()` and I still think we should `return` here (perhaps for future cleanup at exit and avoid the leak).

dagrayvid replied:
Yes, agreed.

dagrayvid replied:
This should now be fixed.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
+
+			klog.V(2).Infof("event from filewatcher on file: %v, event: %v", event.Name, event.Op)
+
+			if event.Op != fsnotify.Chmod && event.Op != fsnotify.Remove &&
```
Agreed it is more readable that way. I had this originally but switched it to save on lines... but on second thought, I think overall the function reads better with it.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -1,25 +1,32 @@
 const (
-	tlsCRT = "/etc/secrets/tls.crt"
-	tlsKey = "/etc/secrets/tls.key"
+	certsDir = "/etc/secrets"
```
I changed it to tlsSecretDir.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
+	return srv
+}
+
+func startServer(srv *http.Server, stopCh <-chan struct{}) {
```
Good catch. This was leftover from my original draft of splitting the runServer function into smaller functions. The stopCh doesn't need to be here as the server will be shutdown if there is a message in stopCh.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
+		klog.Errorf("failed to create watcher: %v", err)
+	}
+	defer watcher.Close()
+	if err := watcher.Add(certsDir); err != nil {
```
This should now be fixed. If the watcher is nil certificate rotation will be "disabled" but the server will still run.

dagrayvid commented:
> Is there a way to add a CI test for this, deleting the secrets twice?
@jmencak @fcami, I spent some time looking into options for testing certificate rotation as part of the e2e test. Originally, I was thinking of using a test pod to curl from the metrics endpoint on the `node-tuning-operator` service, but was unable to get this working using the trusted-ca bundle.... maybe I was doing it wrong :\ 

However, in order to test the full NTO metrics "stack", the idea I'm currently leaning towards is:

1. delete the tls Secret (this will result in [a new certificate and key being put into the secret](https://docs.openshift.com/container-platform/4.9/security/certificates/service-serving-certificate.html#manually-rotate-service-ca_service-serving-certificate)). 
2. wait ~1 minute
3. until success or timeout, query prometheus for an nto metric from the last ~30 seconds and make sure the result is not empty. e.g.:
```
oc exec -n openshift-monitoring -c prometheus prometheus-k8s-0 -- \
curl -s "http://localhost:9090/api/v1/query?query=nto_degraded_info\
&start=$(date -d "30 seconds ago" -u +%FT%T)&end=$(date -u +%FT%T)&step=15s" \
| jq -r
```

If certificate rotation is failing, metrics won't be reported and the result will be something like
```
{
  "status": "success",
  "data": {
    "resultType": "vector",
    "result": []
  }
}
```
If metrics are being scraped from NTO successfully, `data.result` will be nonempty

WDYT? Any better ideas? Maybe I'm missing a simpler way to query prometheus from the e2e test?

fcami commented:
> > Is there a way to add a CI test for this, deleting the secrets twice?
> 
> @jmencak @fcami, I spent some time looking into options for testing certificate rotation as part of the e2e test. Originally, I was thinking of using a test pod to curl from the metrics endpoint on the `node-tuning-operator` service, but was unable to get this working using the trusted-ca bundle.... maybe I was doing it wrong :\
> 
> However, in order to test the full NTO metrics "stack", the idea I'm currently leaning towards is:
> 
> 1. delete the tls Secret (this will result in [a new certificate and key being put into the secret](https://docs.openshift.com/container-platform/4.9/security/certificates/service-serving-certificate.html#manually-rotate-service-ca_service-serving-certificate)).
> 2. wait ~1 minute
> 3. until success or timeout, query prometheus for an nto metric from the last ~30 seconds and make sure the result is not empty. e.g.:
> 
> ```
> oc exec -n openshift-monitoring -c prometheus prometheus-k8s-0 -- \
> curl -s "http://localhost:9090/api/v1/query?query=nto_degraded_info\
> &start=$(date -d "30 seconds ago" -u +%FT%T)&end=$(date -u +%FT%T)&step=15s" \
> | jq -r
> ```
> 
> If certificate rotation is failing, metrics won't be reported and the result will be something like
> 
> ```
> {
>   "status": "success",
>   "data": {
>     "resultType": "vector",
>     "result": []
>   }
> }
> ```
> 
> If metrics are being scraped from NTO successfully, `data.result` will be nonempty
> 
> WDYT? Any better ideas? Maybe I'm missing a simpler way to query prometheus from the e2e test?

Say we delete the secrets twice, we have cert1, cert2 and cert3.
The problem I see is that since cert1 is still valid and not part of any revocation list, the endpoint will continue working in this test even if the operator failed to load cert2 then cert3 and is still using cert1.
Prometheus will still be able to scrape data, so the test is invalid.

The only way to make sure the proper certificate is used seems to be comparing the content of the certificate in the secret with what's exposed / loaded by the operator.

dagrayvid commented:
> The problem I see is that since cert1 is still valid and not part of any revocation list, the endpoint will continue working in this test even if the operator failed to load cert2 then cert3 and is still using cert1.
Prometheus will still be able to scrape data, so the test is invalid.

You're absolutely right, sorry for my misunderstanding.

> The only way to make sure the proper certificate is used seems to be comparing the content of the certificate in the secret with what's exposed / loaded by the operator.

One way to do this is to get the server certificate using the openssl tool:
`openssl s_client -connect node-tuning-operator:60000`. 

Then compare the certificate with the one from the secret.

dagrayvid commented:
@jmencak @fcami, I added an e2e test for certificate rotation. It behaves as expected from my testing. 

I'm not thrilled with the complexity of using Go to exec an oc debug pod in the container from the openshift/tools imagestream, to run an openssl command to get the certs from the server... if you have a more elegant solution I'm open to ideas.

I also am not yet sure why @fcami suggested we should delete the secret twice. From my testing it has the exact same effect as deleting it once.

dagrayvid commented:
/retest 

The failed e2e-aws-operator test was due to infrastructure issues.

dagrayvid commented:
/retest

jmencak commented:
> @jmencak @fcami, I added an e2e test for certificate rotation. It behaves as expected from my testing.
> 
> I'm not thrilled with the complexity of using Go to exec an oc debug pod, using the container from the openshift/tools imagestream, to run an openssl command to get the certs from the server... if you have a more elegant solution I'm open to ideas.

Neither am I "thrilled" about that.  Without digging too much (yet), you might be able to use the `openssl` command on the `/host` from a `TuneD` container.  This works for me on my cluster, but had to use the service IP directly as it wouldn't resolve:
```
/host/usr/bin/openssl s_client -connect 10.130.0.65:60000
```
alternativel, you might be able to use the `crypto/tls` package and the raw structures it provides (again untested):
https://stackoverflow.com/questions/31751764/get-remote-ssl-certificate-in-golang

dagrayvid commented:
Thanks for the ideas Jiri!

> you might be able to use the crypto/tls package and the raw structures it provides (again untested):
https://stackoverflow.com/questions/31751764/get-remote-ssl-certificate-in-golang

I was originally thinking of doing something like this, but the metrics service will only be available from within the cluster, so this go code would need to run in some pod, rather than as part of the e2e test, which makes it a bit uglier IMO.

> you might be able to use the openssl command on the /host from a TuneD container. This works for me on my cluster, but had to use the service IP directly as it wouldn't resolve... Not sure what the best way to resolve the host in the e2e test is, but perhaps just resolving it in the operator pod and then using it in the openshift-tuned pod to issue the openssl command on the host might be the easiest way to solve this?

I think the service IP is just the NTO Pod IP, which we can get by inspecting the NTO Pod object. I will try this.

dagrayvid commented:
@jmencak I added a new version of the e2e test which uses the TuneD pod as discussed.

jmencak commented:
> @jmencak I added a new version of the e2e test which uses the TuneD pod as discussed.

Thank you David, will have another look at this PR soon, likely tomorrow morning.

jmencak commented on test/e2e/basic/metrics_cert_rotation.go:
```diff
@@ -49,15 +60,14 @@ var _ = ginkgo.Describe("[basic][metrics] Node Tuning Operator certificate rotat
-					return false, nil
-				}
-				serverCertContents := stdout.String()
+				ginkgo.By(fmt.Sprintf("Getting certificate from NTO metrics server"))
```
Nit, consistency: `s/Getting/getting/`

dagrayvid replied:
Thanks, fixed this and others.

jmencak commented on test/e2e/basic/metrics_cert_rotation.go:
```diff
@@ -49,15 +60,14 @@ var _ = ginkgo.Describe("[basic][metrics] Node Tuning Operator certificate rotat
-				serverCertContents := stdout.String()
+				ginkgo.By(fmt.Sprintf("Getting certificate from NTO metrics server"))
+				operatorPodIP := operatorPod.Status.PodIP
+				opensslCmd := "/host/usr/bin/openssl s_client -connect " + operatorPodIP + ":60000 2>/dev/null </dev/null |  sed -ne '/-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p'"
```
`s/|  sed/| sed/` # i.e. remove one space.

dagrayvid replied:
Thanks, fixed.

jmencak commented on test/e2e/util/util.go:
```diff
@@ -64,6 +64,26 @@ func GetTunedForNode(cs *framework.ClientSet, node *corev1.Node) (*corev1.Pod, e
+		if len(podList.Items) == 0 {
+			return nil, fmt.Errorf("failed to find the cluster-node-tunig-operator pods")
+		}
+		return nil, fmt.Errorf("too many (%d) cluster-node-tunig-operator pods", len(podList.Items))
```
Do we need to do this?  Say we have two operators running and one waiting to be elected as the leader.  Is this what are you worried about?  Wouldn't it be better to return just the first one?  For what we currently need this would be likely be better and wouldn't fail an e2e test if two operators were running at the same time.

dagrayvid replied:
Yes, this is what I was trying to handle. Taking the first Pod is probably better... though if there are multiple running pods I'm not sure which one the metrics service might point to.

jmencak replied:
Aaah, good point.  I should probably leave the reviews for the morning. :)

jmencak commented on test/e2e/util/util.go:
```diff
@@ -64,6 +64,26 @@ func GetTunedForNode(cs *framework.ClientSet, node *corev1.Node) (*corev1.Pod, e
 	return &podList.Items[0], nil
 }
 
+// GetTunedForNode returns a Pod that runs on a given node.
```
Comment doesn't match what the function does.

dagrayvid replied:
Fixed, sorry about that.

jmencak commented on test/e2e/util/util.go:
```diff
@@ -64,6 +64,26 @@ func GetTunedForNode(cs *framework.ClientSet, node *corev1.Node) (*corev1.Pod, e
+
+	podList, err := cs.Pods(ntoconfig.OperatorNamespace()).List(context.TODO(), listOptions)
+	if err != nil {
+		return nil, fmt.Errorf("couldn't list potential NTO operator pods: %v", err)
```
I *think* it should be `s/pods/Pods/`.  At least that what I use in the logs elsewhere.  Similarly below.

dagrayvid replied:
Fixed in a few spots, thanks.

jmencak commented on test/e2e/util/util.go:
```diff
@@ -64,6 +64,26 @@ func GetTunedForNode(cs *framework.ClientSet, node *corev1.Node) (*corev1.Pod, e
 }
 
+// GetNodeTuningOperator returns the node tuning operator Pod,
+// or the first one found if multiple are running
```
Nit: `s/are running/are running./`

dagrayvid replied:
Fixed this, thanks.

dagrayvid commented:
@jmencak, I tried it with two operator pods running, and the openssl command will work regardless of which NTO pod it is connecting to (as long as they both rotate their certificates). AFAICT, this test is valid even if we don't know which NTO pod we will get, so I removed the check that the pod list is of length 1.

jmencak commented on test/e2e/basic/metrics_cert_rotation.go:
```diff
@@ -0,0 +1,77 @@
+				waitDuration = 5 * time.Minute
+			)
+
+			// Delete the existing secret to force certificate rotation
```
Nit: `s/rotation/rotation./`

dagrayvid replied:
Fixed.

jmencak commented on test/e2e/basic/metrics_cert_rotation.go:
```diff
@@ -0,0 +1,77 @@
+			ginkgo.By("sleeping 5s to wait for Secret to be refreshed")
+			time.Sleep(5 * time.Second)
+
+			// Delete the secret again to force certificate rotation
```
I didn't quite get why we're doing this twice.  Is this based on what @fcami suggested?  Can we have an explanation?

fcami replied:
If the fsnotify watches a file that's deleted, it will report the delete event, which might trigger reloading the certificate properly. However, rotating the certificate a second time will not work since the fsnotify does not watch the new file.
In short, testing that renewing the cert twice results in the last certificate being loaded protects against a refactoring that would simplify the fsnotify logic too much.
Does that help @jmencak ?

jmencak replied:
It does, thank you.  I need to look into this more.  In any case, an explanation like this should go in the comments.

dagrayvid replied:
I added the explanation to the second deletion.

jmencak commented on test/e2e/basic/metrics_cert_rotation.go:
```diff
@@ -0,0 +1,77 @@
+			_, _, err := util.ExecAndLogCommand("oc", "delete", "-n", ntoconfig.OperatorNamespace(), "secret/node-tuning-operator-tls")
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			ginkgo.By("sleeping 5s to wait for Secret to be refreshed")
+			time.Sleep(5 * time.Second)
```
Can we be reasonably sure the secrets are refreshed within in 5 seconds?  Is it possible to do this in a cleaner way without the sleep?  Ignore if it is too hard to implement.

dagrayvid replied:
I rewrote this part into a function that will delete it, then check every 2 seconds for up to 2 minutes whether it is recreated. It seems to usually be back by the second check. I think this is "better" so we don't have to rely on it being refreshed within 5 seconds, but it will also not wait as long most of the time.

(Will update PR once I address the error handling in the server restart)

dagrayvid replied:
Pushed the change to this part.

jmencak reviewed (commented):
Thank you David.  This looks very close to being ready to be merged.  Only a few nits and questions.  I'd also appreciated @fcami input.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
+			}
+
+			if certsChanged(origCertChecksum, origKeyChecksum) {
+
```
Fixed, thanks. It seems I tend to subconsciously add extra whitespace in my code.

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
+// Determine if the certificates have changed and need to be updated.
+// Returns true if both files have changed AND neither is an empty file.
+func certsChanged(origCertChecksum []byte, origKeyChecksum []byte) bool {
+	// Check if both files exist
```
Fixed

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
-// RegisterVersion exposes the Operator build version.
-func RegisterVersion(version string) {
-	buildInfo.WithLabelValues(version).Set(1)
+// Compute the sha256 checksum for a file.
```
Fixed

dagrayvid commented on test/e2e/basic/metrics_cert_rotation.go:
```diff
@@ -0,0 +1,77 @@
+)
+
+var _ = ginkgo.Describe("[basic][metrics] Node Tuning Operator certificate rotation", func() {
+
```
No reason. Fixed.

dagrayvid commented:
I squashed the commits into one. Retitling to include BZ.

/retitle Bug 2026387: Handle certificate rotation in pkg/metrics/server.go

dagrayvid commented on pkg/metrics/server.go:
```diff
@@ -37,44 +44,137 @@ func RunServer(port int, stopCh <-chan struct{}) {
+
+				// restart server
+				klog.Infof("restarting metrics server to rotate certificates")
+				stopServer(srv)
```
As discussed on slack, the alternatives don't seem much better so we will leave it for now.

dagrayvid commented:
/hold

I want to do some final testing, but ready for /lgtm

dagrayvid commented:
/retest

jmencak commented:
/retest

jmencak commented:
As mentioned over chat earlier, the last nit I see is `s/pod if multiple are running/pod if multiple are running./` but I can live with that or can be changed for consistency later.  Thank you again for the PR, it 
/lgtm
now.

jmencak commented:
/approve

dagrayvid commented:
Thanks @jmencak.

/unhold

dagrayvid commented:
/cherry-pick release-4.9
